File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Stack/Recursion/StackRecursion.py
# Recursion in Stack - Recursively pop until empty (MotorPH)


def push(stack, action, brand, quantity):
    """Push inventory action onto stack."""
    stack.append({"action": action, "brand": brand, "quantity": quantity})


def recursive_pop(stack):
    """Recursively pop stack until empty (base case: empty stack)."""
    if not stack:
        return
    item = stack.pop()
    print(f"  Popped: [{item['action']}] {item['brand']}: {item['quantity']} units")
    recursive_pop(stack)


def display(stack):
    """Display stack (top to bottom)."""
    if not stack:
        print("  (stack empty)")
        return
    for i, item in enumerate(reversed(stack)):
        print(f"  {i + 1}. [{item['action']}] {item['brand']}: {item['quantity']} units")


def main():
    # Stack for inventory actions
    stack = []

    # Push inventory actions
    push(stack, "add", "Honda", 5)
    push(stack, "add", "Yamaha", 3)
    push(stack, "add", "Suzuki", 7)

    print("Stack before recursive pop:")
    display(stack)

    # Recursively pop until empty
    print("\nRecursive pop (emptying stack):")
    recursive_pop(stack)

    print("\nStack after recursive pop:")
    display(stack)


if __name__ == "__main__":
    main()
#

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Stack/Recursion/StackRecursion.java
package Stack.Recursion;

import java.util.ArrayList;
import java.util.List;

/**
 * Recursion in Stack - Recursively pop until empty (MotorPH)
 */
public class StackRecursion {

    record StackItem(String action, String brand, int quantity) {}

    /** Push inventory action onto stack. */
    static void push(List<StackItem> stack, String action, String brand, int quantity) {
        stack.add(new StackItem(action, brand, quantity));
    }

    /** Recursively pop stack until empty (base case: empty stack). */
    static void recursivePop(List<StackItem> stack) {
        if (stack.isEmpty()) {
            return;
        }
        StackItem item = stack.remove(stack.size() - 1);
        System.out.println("  Popped: [" + item.action() + "] " + item.brand() + ": " + item.quantity() + " units");
        recursivePop(stack);
    }

    /** Display stack (top to bottom). */
    static void display(List<StackItem> stack) {
        if (stack.isEmpty()) {
            System.out.println("  (stack empty)");
            return;
        }
        for (int i = stack.size() - 1; i >= 0; i--) {
            StackItem item = stack.get(i);
            System.out.println("  " + (stack.size() - i) + ". [" + item.action() + "] " + item.brand() + ": " + item.quantity() + " units");
        }
    }

    public static void main(String[] args) {
        // Stack for inventory actions (ArrayList: add at end = push, remove from end = pop)
        List<StackItem> stack = new ArrayList<>();

        // Push inventory actions
        push(stack, "add", "Honda", 5);
        push(stack, "add", "Yamaha", 3);
        push(stack, "add", "Suzuki", 7);

        System.out.println("Stack before recursive pop:");
        display(stack);

        // Recursively pop until empty
        System.out.println("\nRecursive pop (emptying stack):");
        recursivePop(stack);

        System.out.println("\nStack after recursive pop:");
        display(stack);
    }
}
